src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/failover/FailoverConnectionPlugin.java

```diff
@@ -0,0 +1,1289 @@
+  protected int failoverTimeoutMsSetting;
+  protected int failoverClusterTopologyRefreshRateMsSetting;
+  protected int failoverWriterReconnectIntervalMsSetting;
+  protected int failoverReaderConnectTimeoutMsSetting;
+  protected String clusterIdSetting;
+  protected String clusterInstanceHostPatternSetting;
+  protected int failoverConnectTimeoutMs;
+  protected int failoverSocketTimeoutMs;
+  protected boolean isClosed = false;
+  protected boolean closedExplicitly = false;
+  protected String closedReason = null;
+  // Keep track of the last exception processed in 'dealWithInvocationException()' in order to avoid creating connections repeatedly from each time the same
+  // exception is caught in every proxy instance belonging to the same call stack.
+  protected Throwable lastExceptionDealtWith = null;
+  JdbcConnection parentProxyConnection = null;
+  private long invokeStartTimeMs;
+  private long failoverStartTimeMs;
+  private boolean autoReconnect;
+
+  public FailoverConnectionPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this.currentConnectionProvider = currentConnectionProvider;
+    this.propertySet = propertySet;
+    this.mainHost = this.currentConnectionProvider.getCurrentHostInfo();
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+    this.initialConnectionProps = this.mainHost.getHostProperties();
+
+    initSettings();
+
+    AuroraTopologyService topologyService = new AuroraTopologyService(this.logger);
+    topologyService.setPerformanceMetricsEnabled(this.gatherPerfMetricsSetting);
+    topologyService.setRefreshRate(this.clusterTopologyRefreshRateMsSetting);
+    this.topologyService = topologyService;
+
+    // TODO: need to initialize thisAsConnection, parentProxyConnection and
+    this.connectionProvider = new BasicConnectionProvider();
+    this.readerFailoverHandler =
+        new ClusterAwareReaderFailoverHandler(
+            this.topologyService,
+            this.connectionProvider,
+            this.initialConnectionProps,
+            this.failoverTimeoutMsSetting,
+            this.failoverReaderConnectTimeoutMsSetting,
+            this.logger);
+    this.writerFailoverHandler =
+        new ClusterAwareWriterFailoverHandler(
+            this.topologyService,
+            this.connectionProvider,
+            this.readerFailoverHandler,
+            this.initialConnectionProps,
+            this.failoverTimeoutMsSetting,
+            this.failoverClusterTopologyRefreshRateMsSetting,
+            this.failoverWriterReconnectIntervalMsSetting,
+            this.logger);
+
+    try {
+      initProxy();
+    } catch (SQLException e) {
+      // TODO: review
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public Object execute(
+      Class<?> methodInvokeOn, String methodName, Callable<?> executeSqlFunc)
+      throws Exception {
+    this.invokeStartTimeMs =
```

Comment:
At first, we need to check if failover's enabled. Otherwise, we need just call next plugin.